src/main/scala/apps/cameraPipe.scala

```diff
@@ -0,0 +1,412 @@
+package apps
+
+import rise.core._
+import rise.core.types._
+import rise.core.DSL._
+import rise.core.TypeLevelDSL._
+import rise.core.HighLevelConstructs._
+
+// algorithm taken from Halide (halide-lang.org)
+object cameraPipe {
+  // iN -> iN -> uN
+  val abs_diff: Expr = foreignFun("abs_diff_i16", i16 ->: i16 ->: u16)
+  def clamp(dt: DataType): Expr =
+    foreignFun(s"clamp_${dt}", dt ->: dt ->: dt ->: dt)
+  val max: Expr = foreignFun("max_i16", i16 ->: i16 ->: i16)
+  val pow: Expr = foreignFun("pow_f32", f32 ->: f32 ->: f32)
+
+  def li16(v: Int): Expr = cast(l(v)) :: i16
+  def li32(v: Int): Expr = cast(l(v)) :: i32
+  def lu8(v: Int): Expr = cast(l(v)) :: u8
+
+  // average two positive values rounding up
+  val avg: Expr = dtFun(dt => dtFun(compute_dt => fun(
+    dt ->: dt ->: dt
+  )((a, b) => {
+    val cdt = fun(a => cast(a) :: compute_dt)
+    // FIXME: avoid literal casts?
+    cast((cdt(a) + cdt(b) + cdt(l(1))) / cdt(l(2))) :: dt
+  })))
+
+  case class Image(
+    xBeg: Nat, xEnd: Nat,
+    yBeg: Nat, yEnd: Nat,
+    expr: Expr
+  )
+
+  def letImage(img: Image, f: Image => Expr): Expr =
+    img.expr |> fun(x => f(Image(img.xBeg, img.xEnd, img.yBeg, img.yEnd, x)))
```

Comment:
Is this supposed to be beta reduced during code gen? By the name I assumed that `fun` would be `let`. This will effectively be an `Apply(f, image)`, no?

Reply:
This is not using a `rise.let`. I am open to other naming suggestions. The point is simply to give a name to the image and avoid inlining during expression construction (but not during rewriting and codegen, this has to be decided separaly).

Reply:
I see. I was just wondering. The idea is the same I guess. However, it will not create a `let`, which is what was confusing me. I guess the name that would come to mind right away is `funImage` or `applyFunImage`. But I assume that you chose `letImage` because it is working as a let on the DSL level.
So I guess you should leave the name if your prefer.